src/Nethermind/Nethermind.Consensus/Producers/BlockProducerBase.cs

```diff
@@ -192,7 +192,11 @@ protected bool TrySetState(Hash256? parentStateRoot)
         {
             if (parentStateRoot is not null && StateProvider.HasStateForRoot(parentStateRoot))
             {
-                StateProvider.StateRoot = parentStateRoot;
+                // TODO: do we need this here?
+                using (StateProvider.BeginScope())
+                {
+                    StateProvider.StateRoot = parentStateRoot;
+                }
```

Comment:
This is a pretty weird pattern that emerges

Reply:
Should make StateRoot read only